### ISSUE TITLE
Add support for properties with multiple values

### ISSUE THREAD

**Issue by emersion** [emersion/go-vcard#5]
e.g. `CATEGORIES`

**Comment by latentgenius** [comment]
I'd like to have a go at this. Hopefully I'm up to it.
Also, you can add a **Hacktoberfest** tag to this issue so it shows up in searches for issues with the tag. More info at [https;//hacktoberfest.digitalocean.com/](Hacktoberfest).

**Comment by latentgenius** [comment]
I'd like to have a go at this. Hopefully I'm up to it.
Also, you can add a **Hacktoberfest** tag to this issue so it shows up in searches for issues with the tag. More info at [https://hacktoberfest.digitalocean.com/](Hacktoberfest).

**Comment by emersion** [comment]
Ahah, added the label. :)

This is a bit tricky since some properties might have multiple values, some won't. Also, it's more difficult than just creating one `Field` per value, since parsing a card and then formatting it wouldn't produce the same card as the original one:

```
CATEGORIES: cat 1, cat 2
```

vs

```
CATEGORIES: cat 1
CATEGORIES: cat 2
```

I'm not sure if that matters that much. Maybe reading the relevant parts of the RFC could help to decide.

Also, escaping must be done properly.

**Comment by latentgenius** [comment]
So tricky. That RFC is the most boring thing I've seen so far. I'll try. Maybe there might be insight in other vCard libraries.

**Comment by latentgenius** [comment]
It seems this feature is already present in this library. According to RFC we have to escape comma-separated values in such properties that have multiple values, and in fact, in all properties.
> Some properties may contain one or more values delimited by a COMMA
character (U+002C). Therefore, a COMMA character in a value MUST be
escaped with a BACKSLASH character (U+005C), even for properties that
don’t allow multiple instances (for consistency).

I tried encoding such a field and it resulted in a backslash-escaped commas.
```
CATEGORIES: cat 1\, cat 2
```
But decoding it worked. Importing the vCard in Evolution also worked. Then used the [ExampleNewDecoder](https://github.com/emersion/go-vcard/blob/e65e62c72276de9b5b9b63db2ffe62fb28d8f48d/example_test.go#L12-L30) to decode the card, replacing the preferred value thus
```diff
- log.Println(card.PreferredValue(vcard.FieldFormattedName))
+ log.Println(card.PreferredValue(vcard.FieldCategories))
```
The output was
```
CATEGORIES: cat 1, cat 2
```

**Comment by emersion** [comment]
Yeah, this is done by https://github.com/emersion/go-vcard/blob/master/encoder.go#L74 and https://github.com/emersion/go-vcard/blob/master/decoder.go#L224. But this issue is not about fields with a single value containing a `,` character, but rather about fields with multiple values (separated by non-escaped `,`).

**Comment by latentgenius** [comment]
Does this imply that the comma in `CATEGORIES: cat 1, cat 2` should not be escaped, since it serves as a separator for the field's values?

**Comment by emersion** [comment]
No. It means that for now, users cannot put multiple values in one field. They have to create two fields:

```
CATEGORIES: cat 1
CATEGORIES: cat 2
```

Also, as the RFC says, all commas must be escaped in field values (`CATEGORIES` and `NICKNAME` are not special cases).

**Comment by akshaychhajed** [comment]
I think we are interpreting incorrectly.

CATEGORIES: cat1, cat2 

> Some properties may contain one or more values delimited by a COMMA
> character (U+002C). Therefore, a COMMA character **_in a value_** MUST be
> escaped with a BACKSLASH character (U+005C), even for properties that
> don’t allow multiple instances (for consistency).

This statement talks about value in multiple value context.

i.e
it should escape "," if value is "cat1, cat2" (single value with "," within)
It should not escape if values are "cat1" and "cat2". (multiple values where "," acts as seperator)


The comment talks about escaping "," within values and not when it acts as separator.
I think @latentgenius is correct in assuming NICKNAME and CATEGORIES are special cases due to their text-list type.

**Comment by oliverpool** [comment]
#35 is an attempt to address this (quite old) issue.

**Comment by lpgonzalez** [comment]
Following up on this old issue, I'd like to fix the specific list-separator case for `CATEGORIES` and `NICKNAME`, and I think it fits the "dedicated getter/setter" direction you mentioned in #35.

I ran into this serving vCards over CardDAV: `SetCategories([]string{"A", "B"})` encodes as `CATEGORIES:A\,B`, and conformant clients (DAVx5, Evolution) then show the whole list as a single category `"A,B"` instead of two.

Agreement first: I fully agree that a comma inside a text value MUST be escaped (RFC 6350 §3.4: "even for properties that don't allow multiple instances, for consistency"). I'm not proposing to change `formatValue` or its test, and TEXT properties like `NOTE`/`FN` keep emitting `\,`. The change is scoped to exactly the two properties the RFC defines as comma-separated lists: `CATEGORIES` (§6.7.1) and `NICKNAME` (§6.2.3). For those, the comma between items is a separator, so escaping it merges the list:

```go
card.SetCategories([]string{"A", "B"})
// today encodes: CATEGORIES:A\,B   → a conformant parser reads ONE item "A,B"
// should encode:  CATEGORIES:A,B
```

Why a getter/setter alone isn't enough (the tricky bit): the decoder is lossy. `parseValue` unescapes the whole value, so both inputs below decode to the same `Field.Value` `"A,B"`:

```
CATEGORIES:A,B    → Field.Value = "A,B"   (two items)
CATEGORIES:A\,B   → Field.Value = "A,B"   (one item "A,B")
```

So a getter/setter working on the post-decoded string can't recover the distinction or round-trip correctly without the codec's cooperation. The minimal fix is to skip comma (un)escaping for just these two registered properties, exactly like `;` is already left untouched for `N`/`ADR`/`GENDER` and handled by their typed `Name`/`SetName`, `Gender`/`SetGender` accessors; no new exported type (unlike #35), just the existing `Categories`/`SetCategories` and new `Nicknames`/`SetNicknames`. Commas inside an item stay escaped:

```go
card.SetCategories([]string{"Big, Tall", "Small"})
// encodes: CATEGORIES:Big\, Tall,Small   (escaped comma = literal, bare = separator)
```

Representation - your call. I'd lead with the single-line canonical form (`CATEGORIES:A,B`), which keeps one property instance and matches what most cards in the wild use. But if you prefer the multi-instance model you suggested back in 2017 (`CATEGORIES:A` / `CATEGORIES:B`, one value per `Field`), that also works and avoids any codec change - happy to go that way instead. Which do you prefer?

I already have a failing test that reproduces this; the patch itself is small and I'll open a PR as soon as you pick the representation.

**Comment by emersion** [comment]
How about the following plan?

1. Fix the current `Categories`/`SetCategories` to accept/produce multiple fields (one per category). This can be made without breaking API.
2. Make it so unescaped comma-separated values are parsed as multiple `Field`s. This doesn't reflect the original vCard structure, but should be semantically equivalent, I think. Again, this does not break API.
3. Consider replacing `Field.Value` with `Field.Values` containing a `[]string`. This would better match the RFC, but would be an invasive breaking change for downstream users. Moreover, most properties use a single value. Not sure it's worth the trouble. (Alternatively, we could avoid exposing a raw string, and only provide type-safe helpers to manipulate the value.)

**Comment by lpgonzalez** [comment]
Thanks for taking a look. The plan makes sense, and I like that steps 1 and 2 land the fix without breaking the API.

One thing worth scoping carefully in step 2: splitting unescaped commas into multiple `Field`s is right for comma-list properties (`CATEGORIES`, `NICKNAME`), but it would corrupt the semicolon-structured ones (`N`, `ADR`, `ORG`, ...), where a top-level comma is a component-internal value separator rather than a property-value separator - e.g. `N:Public;John,J.;;;` or `ADR:;;1 Main St,Apt 2;…`. So the split probably needs to be gated by the property's value type (the comma-list properties / the `VALUE` parameter) rather than applied to every value; otherwise a round-trip of a structured property loses its shape.

Step 3 (`Field.Values []string`) is the RFC-pure model, but I agree the breaking change isn't really justified on its own. 1+2 already remove the data loss for downstream users. If you ever cut a major, the type-safe-helpers route you mention is appealing.

Happy to help land this: I can put up a PR, and/or contribute round-trip fixtures - both the list-valued cases (`CATEGORIES`/`NICKNAME`) and a few guard cases pinning that `N`/`ADR`/`ORG` keep their structure. I have a small adversarial vCard corpus from a CardDAV server I'm building that hits exactly these. Let me know what would be most useful.

**Comment by emersion** [comment]
Hm, I feared something like this. I'm a bit confused though: [section 6.2.2](https://datatracker.ietf.org/doc/html/rfc6350#section-6.2.2) says "Value type:  A single structured text value.  Each component can have multiple values.". However, the `text` ABNF doesn't allow for unescaped commas.

It seems like multiple values handling has been a common issue with the spec: https://errata.rfc-editor.org/eid3488/

So yeah, in the end it seems like we'd better keep an allow-list of keys whose values are separated by commas and leave the rest untouched.

**Comment by lpgonzalez** [comment]
Agreed, and RFC 6350 actually pins the allow-list down for us, so it's not a
heuristic: comma-separated multiple values are only permitted in N, NICKNAME, ADR,
and CATEGORIES. Everywhere else an unescaped comma is literal, never a separator.

One subtlety worth baking in so we don't over-split: of those four, N and ADR are
*compound* (§3.4 notes they're semicolon-delimited structured values), and the
comma-list there lives *inside* a component, after the semicolon split. So a
top-level "split the whole value on unescaped commas" is only correct for NICKNAME
and CATEGORIES (flat text-lists). For N/ADR, splitting the raw value on commas would
scramble the structure; their comma-multivalue is per-field, not top-level. (That's
also what resolves the §6.2.2 confusion: N's "each component can have multiple values"
is real, but it's per component, inside the ';' structure.)

So I'd scope step 2 as:
- top-level comma, multiple Fields: NICKNAME, CATEGORIES only;
- N, ADR: keep the existing semicolon-structured handling; the per-component
  comma-list is a separate concern (arguably out of scope for the flat Field model);
- everything else (incl. ORG, X-*, unknown IANA): left untouched.

Happy to send a PR for that, plus round-trip and non-corruption fixtures from a
real-world corpus I maintain, proving NICKNAME/CATEGORIES split correctly while
N/ADR/ORG/X- round-trip byte-identical. Want me to open it?

**Comment by emersion** [comment]
Yup, that sounds good to me!

**Comment by lpgonzalez** [comment]
Hi @emersion, I'd like to revisit comma-separated list handling for `CATEGORIES` and
`NICKNAME`. I have a working implementation along the lines you preferred in #35
(dedicated getters/setters, no new value type) and wanted to check the scope with you
before opening a PR.

Scope (intentionally narrow). Top-level comma splitting applies only to the two
properties RFC 6350 defines as comma-separated text lists: `NICKNAME` (§6.2.3) and
`CATEGORIES` (§6.7.1). Everything else is untouched. `N` and `ADR` keep their commas
(they're `;`-structured, so a comma lives inside a component), and `ORG`, `X-*` and
unknown/IANA properties are left exactly as today. An escaped comma (`\,`) never splits,
and only the ASCII comma `U+002C` separates (look-alike Unicode commas such as `U+3001`
/`U+FF0C` don't).

Non-breaking. A comma-separated line decodes into one `*Field` per value, the per-field
model you suggested in this thread back in 2017, so there's no new `Field.Values` API and
the encoder is unchanged (it already emits one line per field). `Categories()` and
`SetCategories()` keep their signatures (now backed by multiple fields), and I added
matching `Nicknames()`/`SetNicknames()`.

Testing, focused on not corrupting anything:
- Byte-for-byte round-trip of structured/other properties with escaped internal commas;
  for a bare comma inside a structured value it stays a single field and re-encoding is
  stable (idempotent).
- An escape-aware split battery (odd/even backslash counts; empty, leading and trailing
  items).
- i18n: multibyte values preserved per item, plus a combining diacritic (`e + U+0301`)
  that must round-trip without NFC normalization.

Covered by unit tests plus an end-to-end pass over a fixture corpus; all green.

Would you prefer I open a PR for this? Happy to show you the diff first if that's easier
to review.

Thanks!
